nntrainer/layers/conv2d_layer.cpp

```diff
@@ -243,59 +295,17 @@ void Conv2DLayer::calcDerivative() {
    *       x [(input_dim.height() + padding[0]*2)
    *           *(input_dim.width() + padding[1]*2) (width)]
    */
-  using uint = unsigned int;
-
-  uint kernel_total_size = kernel_size[0] * kernel_size[1];
-
-  Tensor imKernel(1, 1, in_dim.channel(), filter_size * kernel_total_size);
-  float *imKernel_raw = imKernel.getData();
-
-  for (uint channel_idx = 0; channel_idx < in_dim.channel(); ++channel_idx) {
-    /// each row contains all kernel element in particular channel.
-    uint row_size = kernel_total_size * filter_size;
-    for (uint filter_idx = 0; filter_idx < filter_size; ++filter_idx) {
-
-      /// starting index of each kernel in imKernel
-      float *start =
-        imKernel_raw + channel_idx * row_size + filter_idx * kernel_total_size;
-      /// starting index of each channel in filter
-      float *filter_start =
-        filter_kernel.getAddress(filter_idx, channel_idx, 0, 0);
-
-      std::reverse_copy(filter_start, filter_start + kernel_total_size, start);
-    }
-  }
-
-  TensorDim kdim(in_dim.channel(), filter_size, kernel_size[0], kernel_size[1]);
-
-  TensorDim ret_dim{in_dim.channel(), in_dim.height() + padding[0] * 2,
-                    in_dim.width() + padding[1] * 2};
-
-  TensorDim ret_dim_squeezed{
-    {ret_dim.channel(), ret_dim.height() * ret_dim.width()}};
+  Tensor flipped = filter_kernel.convFlip();
 
-  /// @todo: optimize this as suggested in #825
-  /// we will need to have a zero cost image view by manipulating stride
-  Tensor ret = Tensor(ret_dim_squeezed);
-
-  Tensor im2col_result;
-
-  for (unsigned int b = 0; b < in_dim.batch(); ++b) {
-    Tensor inSub = derivative.getBatchSlice(b, 1);
-
-    im2col(inSub, kdim, same_pad, stride, true, im2col_result);
-
-    ret.reshape(ret_dim_squeezed);
-    imKernel.dot(im2col_result, ret, false, true);
-    ret.reshape(ret_dim);
+  std::array<unsigned int, CONV2D_DIM> same_pad;
+  same_pad[0] = kernel_size[0] - 1;
+  same_pad[1] = kernel_size[1] - 1;
 
-    strip_pad(ret, padding.data(), net_input[0]->getGradientRef(), b);
-  }
+  conv2d_op<ChannelToCol>(derivative, flipped, net_input[0]->getGradientRef(),
```

Comment:
the stride passed as argument here is independent of the input stride?

Reply:
It depends, #929 resolves the issue :)

Reply:
#955 resolves this issue :)